test/integration/base_integration_test.cc

```diff
@@ -694,4 +698,45 @@ AssertionResult BaseIntegrationTest::compareDeltaDiscoveryRequest(
   return AssertionSuccess();
 }
 
+void BaseIntegrationTest::checkForMissingTagExtractionRules() {
+  BufferingStreamDecoderPtr response = IntegrationUtil::makeSingleRequest(
+      test_server_->adminAddress(), "GET", "/config_dump", "", Http::CodecType::HTTP1);
+  ASSERT_TRUE(response->complete());
+  Json::ObjectSharedPtr json = Json::Factory::loadFromString(response->body());
+  std::vector<std::string> stat_prefixes;
+  Json::ObjectCallback find_stat_prefix = [&](const std::string& name,
+                                              const Json::Object& root) -> bool {
+    if (name == "stat_prefix") {
+      auto prefix = root.asString();
+      if (!prefix.empty()) {
+        stat_prefixes.push_back(prefix);
+      }
+    } else if (root.isObject()) {
+      root.iterate(find_stat_prefix);
+    } else if (root.isArray()) {
+      std::vector<Json::ObjectSharedPtr> elements = root.asObjectArray();
+      for (const auto& element : elements) {
+        find_stat_prefix("", *element);
+      }
+    }
+    return true;
+  };
+  find_stat_prefix("", *json);
+  ENVOY_LOG_MISC(debug, "discovered stat_prefixes {}", stat_prefixes);
+
+  auto check_metric_array = [&](auto vec) {
+    for (const auto& metric : vec) {
+      const std::string tag_extracted_name = metric->tagExtractedName();
+      for (const std::string& stat_prefix : stat_prefixes) {
+        EXPECT_EQ(tag_extracted_name.find(stat_prefix), std::string::npos)
+            << "Missing stat tag-extraction rule for stat '" << tag_extracted_name
+            << "' and stat_prefix '" << stat_prefix << "'";
+      }
+    }
+  };
+  check_metric_array(test_server_->counters());
```

Comment:
nit; could use test_server_->forEachCounter here and void making the intermediate array. Not sure if that would be material.